java-checks/src/main/java/org/sonar/java/checks/tests/ParameterizedTestCheck.java

```diff
@@ -0,0 +1,160 @@
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.java.model.JavaTree;
+import org.sonar.java.model.SyntacticEquivalence;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.SymbolMetadata;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeCastTree;
+
+import static java.util.Arrays.asList;
+
+@Rule(key = "S5976")
+public class ParameterizedTestCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Replace these %d tests with a single Parameterized one.";
+
+  private static final Set<String> TEST_ANNOTATIONS = new HashSet<>(asList(
+    "org.junit.Test",
+    "org.junit.jupiter.api.Test",
+    "org.testng.annotations.Test"));
+
+  private static final int MIN_SIMILAR_METHODS = 3;
+  private static final int MIN_NUMBER_STATEMENTS = 2;
+  private static final int MAX_NUMBER_PARAMETER = 3;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    List<MethodTree> methods = classTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .filter(ParameterizedTestCheck::isParametrizedCandidate)
+      .collect(Collectors.toList());
+    if (methods.size() < MIN_SIMILAR_METHODS) {
+      return;
+    }
+
+    Set<MethodTree> handled = new HashSet<>();
+    for (int i = 0; i < methods.size(); i++) {
+      MethodTree method = methods.get(i);
+      List<StatementTree> methodBody = method.block().body();
+      // In addition to filtering literals, we want to count the number of differences since they will represent the number of parameter
+      // that would be required to transform the tests to a single parametrized one.
+      CollectAndFilter collectAndFilter = new CollectAndFilter();
+
+      List<MethodTree> equivalentMethods = methods.stream()
+        .skip(i + 1L)
+        // avoid reporting multiple times
+        .filter(otherMethod -> !handled.contains(otherMethod))
+        // only consider method syntactically equivalent, ignoring literals
+        .filter(otherMethod -> {
+          boolean areEquivalent = SyntacticEquivalence.areEquivalent(methodBody, otherMethod.block().body(), collectAndFilter);
+          if (areEquivalent) {
+            // If methods where not equivalent, we don't want to pollute the set of node to parameterize.
+            collectAndFilter.validateCollection();
```

Comment:
I don't understand how this code can accidentally work.
- If a method is equivalent, we copy `currentNodeToParameterize` into `nodeToParametrize`, without cleaning `currentNodeToParameterize`, it's not optimum
- If a method is not equivalent, we do nothing! we keep `currentNodeToParameterize` dirty and it's content will be copied into `nodeToParametrize` by the following equivalent method :confounded: